### PR TITLE
Optimize on replica send order

### PR DESCRIPTION
**Pull request by thundercw**
If there is a replica in local node, move it to the front.

### CODE REVIEW

**Comment by spencerkimball** [comment]
LGTM with exception of adding the leaktest to the new unittest:

	defer leaktest.AfterTest(t)

**Comment by thundercw** [comment]
@spencerkimball  updated , PTAL !

**Comment by spencerkimball** [comment]
LGTM

**Comment by tschottdorf** [comment]
That means that if a local replica doesn't know the current leader, the `DistSender` will not try to talk to anyone else. That should be rare though so LGTM.

**Comment by tamird** [comment]
> That means that if a local replica doesn't know the current leader, the DistSender will not try to talk to anyone else.

I don't follow, why is that the case?

File: kv/dist_sender_test.go [hunk 1]
```diff
@@ -90,6 +90,62 @@ func makeTestGossip(t *testing.T) (*gossip.Gossip, func()) {
...
+			},
+			localNodeDesc: proto.NodeDescriptor{NodeID: 1, Attrs: proto.Attributes{Attrs: []string{"ab"}}},
+		},
+	}
+	for _, test := range testCase {
+		ctx := &DistSenderContext{nodeDescriptor: &test.localNodeDesc}
+		ds := NewDistSender(ctx, nil)
+		ds.optimizeReplicaOrder(test.slice)
+		if s := test.slice[0]; s.NodeID != ctx.nodeDescriptor.NodeID {
+			t.Errorf("The local replica is not move to front.")
```
Comment (tamird) [inline kv/dist_sender_test.go]: nit: this error message could be better (i.e. it should log all the node IDs)

**Comment by tschottdorf** [comment]
> > That means that if a local replica doesn't know the current leader, the DistSender will not try to talk to anyone else.

> I don't follow, why is that the case?

* local replica moved to front
* send (stable order) returns NotLeaderError without hint
* local replica moved to front
* send (stable order) returns NotLeaderError without hint
* ... (until replica knows a leader)

**Comment by tamird** [comment]
> - local replica moved to front
- send (stable order) returns NotLeaderError without hint
- local replica moved to front

I think `Send` would just move on to the next replica in this case.

**Comment by tschottdorf** [comment]
no, as far as `rpc.Send` is concerned, the request is successful. `Send` only cares about `rpc` errors, and there aren't any.

**Comment by tamird** [comment]
Ah, yes.

**Comment by thundercw** [comment]
@tamird Updated, PTAL .thanks.

**Comment by thundercw** [comment]
@tschottdorf Yes,that will be a problem. I think in most time it will get the leader from NotLeaderError response. Send to local replica will improve the performance specially the replicas are located in different DC . If our home-grown RPC supports calling local functions directly, it will skip the network.

File: kv/dist_sender_test.go [hunk 2]
```diff
@@ -90,6 +90,62 @@ func makeTestGossip(t *testing.T) (*gossip.Gossip, func()) {
...
+			},
+			localNodeDesc: proto.NodeDescriptor{NodeID: 1, Attrs: proto.Attributes{Attrs: []string{"ab"}}},
+		},
+	}
+	for _, test := range testCase {
+		ctx := &DistSenderContext{nodeDescriptor: &test.localNodeDesc}
+		ds := NewDistSender(ctx, nil)
+		ds.optimizeReplicaOrder(test.slice)
+		if s := test.slice[0]; s.NodeID != ctx.nodeDescriptor.NodeID {
+			t.Errorf("unexpected header , wanted nodeid = %d, got %d ", ctx.nodeDescriptor.NodeID, s.NodeID)
```
Comment (tamird) [inline kv/dist_sender_test.go]: extra space here before the comma

**Comment by tamird** [comment]
LGTM

**Comment by thundercw** [comment]
@tamird  Updated.

**Comment by tamird** [comment]
Thanks!

**Pull request merged**